src/conn_poll.c

```diff
@@ -141,299 +313,487 @@ int conn_poll_prepare(conn_registry_t *registry, pfds_record_t *pfds, timestamp_
+	return 0;
 }
 
-int conn_poll_recv_udp(socket_t sock, char *buffer, size_t size, addr_record_t *src) {
-	JLOG_VERBOSE("Receiving datagram");
-	int len;
-	while ((len = udp_recvfrom(sock, buffer, size, src)) == 0) {
-		// Empty datagram, ignore
-	}
+static bool conn_poll_mark_tcp_connection_closed(juice_agent_t *agent, conn_impl_t *conn_impl,
+	                                              size_t index) {
+	if (index >= conn_impl->tcp_conns_count)
+		return false;
 
-	if (len < 0) {
-		if (sockerrno != SEAGAIN && sockerrno != SEWOULDBLOCK)
-			JLOG_ERROR("recvfrom failed, errno=%d", sockerrno);
+	addr_record_t remote = conn_impl->tcp_conns[index].remote;
+	conn_poll_free_tcp_connection(conn_impl->tcp_conns + index);
+	return agent_conn_tcp_state(agent, &remote, TCP_STATE_DISCONNECTED) == 0;
+}
 
-		return -sockerrno;
+static int conn_poll_tcp_flush(tcp_connection_t *connection) {
+	if (connection->connecting)
+		return 0;
+	while (connection->send_offset < connection->send_size) {
+		int ret = (int)send(connection->sock,
+		                  (const char *)connection->send_buffer + connection->send_offset,
+		                  (int)(connection->send_size - connection->send_offset), TCP_SEND_FLAGS);
+		if (ret < 0) {
+			if (sockerrno == SEAGAIN || sockerrno == SEWOULDBLOCK)
+				return 0;
+			JLOG_WARN("TCP send failed, errno=%d", sockerrno);
+			return -1;
+		}
+		if (ret == 0)
+			return -1;
+		connection->send_offset += (size_t)ret;
 	}
 
-	addr_unmap_inet6_v4mapped((struct sockaddr *)&src->addr, &src->len);
-	return len; // len > 0
+	connection->send_offset = 0;
+	connection->send_size = 0;
+	return 0;
 }
 
-void conn_poll_process_tcp(juice_agent_t *agent, struct pollfd *pfd) {
-	conn_impl_t *conn_impl = agent->conn_impl;
+static int conn_poll_tcp_queue_send(tcp_connection_t *connection, const char *data, size_t size) {
+	if (size > 65535) {
+		JLOG_WARN("Trying to send TCP frame larger than 65535 bytes");
+		return -1;
+	}
+	if (connection->closed)
+		return -1;
 
-	if (pfd->revents & POLLNVAL) {
-		JLOG_WARN("Invalid socket");
-		return;
+	size_t pending = connection->send_size - connection->send_offset;
+	if (pending > 0 && connection->send_offset > 0) {
+		memmove(connection->send_buffer, connection->send_buffer + connection->send_offset, pending);
+		connection->send_size = pending;
+		connection->send_offset = 0;
+	} else if (pending == 0) {
+		connection->send_size = 0;
+		connection->send_offset = 0;
 	}
 
-	if (pfd->revents & POLLERR || (pfd->revents & POLLHUP && !(pfd->revents & POLLIN))) {
-		JLOG_DEBUG("TCP connection failed");
-		conn_poll_change_tcp_fail(agent);
-		return;
+	size_t frame_size = size + 2;
+	size_t required = connection->send_size + frame_size;
+	if (required > connection->send_capacity) {
+		size_t new_capacity = connection->send_capacity ? connection->send_capacity : 512;
+		while (new_capacity < required)
+			new_capacity *= 2;
+		uint8_t *new_buffer = realloc(connection->send_buffer, new_capacity);
```

Comment:
You must not reallocate to fit everything. If the user sends faster than network path capacity, memory consumption will grow unchecked. This flaw would make it extremely easy for a client to crash a media server by simply starting fixed-rate video flows over TCP and stop reading from its TCP sockets.